### ISSUE TITLE
GRPC C++ Async performance

### ISSUE THREAD

**Issue by nscsekhar** [grpc/grpc#6504]
Hi,

I am measuring GRPC (C++) performance to consider it for a heavily chatty application. 

My environment is something like this:

1. Both Client and Server are processes running on a 1.3GHz server with 2 cores.
2. Client and Server talk asynchronously over Unix Domain sockets

I measured the performance using basic HelloWorld async calls, where Client sends 10 rpcs before waiting for all 10 responses. I measured the time for sending 1 million calls and could get up to 25K rpcs per second.

Is 25K/sec the max I can get? Do we have any performance data for GRPC?

I am looking for atleast 100K rpcs per second in the environment described above.

Thanks,
Surya

**Comment by nscsekhar** [comment]
Attaching the snippets of server and client used to test the performance.
[client-snippet.txt](https://github.com/grpc/grpc/files/256344/client-snippet.txt)
[server-snippet.txt](https://github.com/grpc/grpc/files/256345/server-snippet.txt)

**Comment by vjpai** [comment]
There's no intrinsic limit; I'll look at your snippets to offer suggestions.

**Comment by vjpai** [comment]
Ok, the biggest problem in your code right now is that you're only registering one RPC at a time on the server-side. You should actually register lots of them at once; a good example of this is how we do it in the QPS test. If an incoming RPC matches a pre-registered RPC when it reaches the server, it goes on a fast path but otherwise it ends up on a slow path.

**Comment by vjpai** [comment]
The second issue is something that's missing from your snippets - I don't see where you're actually calling cq.Next . I presume that you're doing that somewhere? But unlike greeter_async_server.cc and greeter_async_client.cc, I would suggest that you have each process have 2 threads and have each of them do cq.Next in each thread. The reason is that your workflow has the client sending a bunch of RPCs, then the server picking up a bunch of RPCs, then the client picking up a bunch of responses. Bursty traffic like that would benefit from having enough threads to support the burst load, not just the steady-state load.

**Comment by vjpai** [comment]
And a third issue that I would bring up is possible client-side limitation. You have 10 RPCs outstanding a time and then wait on all of them. I would suggest two things here:

1. 10->100 or some larger number. The async API's purpose is to push/process a lot of RPCs on a channel from a single thread, so go ahead and try to fill the channel. 
2. Don't wait for all your RPCs to return; just send your next one as soon as one returns. Make the RPCs independent of each other. Otherwise you're counting the stall time waiting on load-balance issues rather than just waiting on the RPCs.

**Comment by nscsekhar** [comment]
Hi Vijay,

Thanks for your suggestions. I tried these by looking at test/cpp/qps - server_async.cc and client_async.cc. (Attached the files.)

1. Client operates as two threads : sender of rpcs and collector of the responses.
2. Server operates as five threads, all working off same socket. 

With this, I see that client is able to push rpcs faster at 100K per second. However, it comes with a set of limitations.

1. Given that I have only one core for Client and one core for Server - multithreading actually degrades the overall performance. The total time taken, i.e. when the collector is done, has now gone up to 170 seconds. In my previous attempt, it was 86 seconds.

2. I had to pause the collector till the sender is done with all RPCs, otherwise sharing/locking the rpc state is degrading the performance quite a bit.

3. The amount of transient memory for the outstanding rpcs (i.e. from the sender allocating it and the collector freeing it) is high. For 1M RPCs, it took 5GB. 

4. More on memory management - the GRPC Async calls did not allow to pass smart pointers. And I could not free the ClientContext in the context of collector thread, as ClientContext is referenced as a private member of the ClientAsyncResponseReader which the collector has.
 
Can you please take a look at my test code and suggest if I try anything else.

One question - Can we have RPCs that dont need a response, so that we can skip the wait overhead. Something like a void return rpc. Today, protoc doesn’t allow specifying such as rpc. If required, the end user can define another RPC to verify if the previous RPCs really did the job.

Thanks,
Surya









[PerfTest.proto.txt](https://github.com/grpc/grpc/files/264060/PerfTest.proto.txt)
[PerfTestClient.cpp.txt](https://github.com/grpc/grpc/files/264059/PerfTestClient.cpp.txt)
[PerfTestServer.cpp.txt](https://github.com/grpc/grpc/files/264061/PerfTestServer.cpp.txt)

**Comment by vjpai** [comment]
Hi there, I'm glad this was able to push your performance higher. On your topics:

1. If performance is better with single-threading, then no problem, use single-threading. I can definitely see that this would be a very platform-specific parameter, and having only 2 cores can lead to very different design choices than having more cores.
2. By all RPCs, do you mean all 1 million? That shouldn't be required.
3. Again, do you have 1 million RPCs outstanding at the same time?
4. The Async API essentially assumes that you know what you're doing. You can free the ClientContext when the RPC is actually correctly Finish'ed . Until then there is always the chance that the ClientContext will still be needed

I will try to look at the test code some time in the next few days. Best of luck and please keep us posted!

**Comment by vjpai** [comment]
One more thing: for one-directional RPCs, you can get this effect by using a streaming RPC rather than a unary RPC.

**Comment by nscsekhar** [comment]
Thanks.. I will try the streaming RPCs first.

**Comment by vjpai** [comment]
I would highly recommend against having 1 million RPCs outstanding at the same time also. This is technically legal but will not be good for latency. 1000 would be very reasonable.

**Comment by nscsekhar** [comment]
Yeah, and not a practical scenario too. I did that just to measure the max performance to push the RPCs before having to collect the responses. With streaming, I can keep it more practical - like stream a 100 messages per RPC. 

My previous attempt with async streaming didnt go well. The examples I looked at were more of ping-pong type. When I ran a loop to just write from client and just receive on the server, both of them asserted. Will spend more time on it now.

**Comment by nscsekhar** [comment]
I could test the async streaming and the performance looks great now. I could see a best case of around 200K RPCs/sec

I measured the total time taken for 1 Million RPC requests sent from client and the server to send 1 response for N requests, where N is 1,2,5,10,20,100,1000. Here are the numbers I observed:

Reply for every Request      : 15K /sec
Reply for every 2 Requests : 34K /sec
Reply for every 5 Requests : 66K /sec
Reply for every 10 Requests : 71K /sec
Reply for every 20 Requests : 111K /sec
Reply for every 100 Requests : 166K /sec
Reply for every 1000 Requests : 166K ~ 200K/sec

Thanks again for your help!
